Bug 4253: ssl_bump prevents from accessing some web contents

chtsanti opened the pull request:
Surrogate-Capability should only be sent on accel traffic, not
on bumped traffic.

This is a Measurement Factory project

chtsanti commented:
This is also fixes bug 4880 (it is duplicate of 4253)

yadij commented:
Can we not move the switchedToHttps() exceptional code to prepareTransparentURL() and use that far simpler function?

rousskov commented:
I assume Amos is suggesting, for switchedToHttps() requests, to call prepareTransparentURL() instead of calling prepareAcceleratedURL(). I do not know whether both functions would produce identical URLs in SslBump context, but, when answering Amos question, please keep in mind that only prepareAcceleratedURL() sets flags.accel, and there may still be SslBump code that relies on that flag to be set.

chtsanti commented:
> Can we not move the switchedToHttps() exceptional code to prepareTransparentURL() and use that far simpler function?

Well, these are the questions I wanted to avoid when I proposed an one line change solution!
Now I am a little lost in cases.

The sslBumped requests are not accelarator requests but they are not exactly transparent, so neither the prepareTransparentURL can be used.
Actually the prepareAcceleratedURL builds correct URL for sslBump only if the Host header exist. The prepareTransparentURL will produce correct URL only if Host header exist to.

I conclude to make a new function prepareTlsSwitchingURL. If the port field is missing from URL in some cases is it a huge problem?  Probably can retrieved if required, just may require a little code.

rousskov commented:
> Amos: Can we not move the switchedToHttps() exceptional code to prepareTransparentURL() and use that far simpler function?

> Christos: these are the questions I wanted to avoid when I proposed an one line change solution!

I am confused by the _combination_ of your reaction to that question and your proposal to add prepareTlsSwitchingURL(): AFAICT, the short answer to Amos' question is "yes" if prepareTransparentURL() does what SslBump needs. The short answer is "no" if prepareTransparentURL() does not do what SslBump needs. And if we cannot really tell what SslBump needs exactly because there are many complex and/or unknown cases, then writing prepareTlsSwitchingURL() to do what SslBump needs (as you propose) sounds like a rather dangerous path!

> I conclude to make a new function prepareTlsSwitchingURL. 

In general, writing prepareTlsSwitchingURL() could be a very good first step. If the result looks very similar to one of the two existing functions mentioned above, then we can avoid code duplication by merging new and old code. If all three functions are very different, then the needs of their callers are probably very different, and all three functions deserve to co-exist (they can still reuse common code, of course). However, if we do not know what prepareTlsSwitchingURL() should do, then this lets-write-and-then-decide logic is not really applicable because we cannot write that function.

BTW, how does the current "Surrogate-Capability should not be sent on bumped traffic" PR description relate to the current PR code changes? Does `flags.accel` control nothing but sending Surrogate-Capability?

yadij commented:
@rousskov, setting flags.accel enables the logic for handling Surrogate-Control. The bug 4880 is a server brokenly emitting that header without Squid sending the Surrogate-Capability header. Both bugs and several others are separate side effects of the flag being set.

yadij commented:
@christos, I can go with having a third function for now. But I do expect that you will find it to be almost identical to the existing transparent one.

chtsanti commented:
> > Christos: these are the questions I wanted to avoid when I proposed an one line change solution!
> 
> I am confused by the _combination_ of your reaction to that question ...

OK, I was tried to avoid looking more on this issue. The two prepare*() functions, even if it is clear what they do, they looked a little complex with the first look!

> 
> BTW, how does the current "Surrogate-Capability should not be sent on bumped traffic" PR description relate to the current PR code changes? Does `flags.accel` control nothing but sending Surrogate-Capability?

My sense is that if the flags.accel flag is not set then squid will not try to process surrogate-control replies. But I did not test.

> @yadij , setting flags.accel enables the logic for handling Surrogate-Control. The bug 4880 is a server brokenly emitting that header without Squid sending the Surrogate-Capability header. Both bugs and several others are separate side effects of the flag being set.

I am not sure that the reported server is broken. In my tests, the site appeared as broken did not enable esi when I am not sending the Surrogate-Capability header.


>  @yadij   I can go with having a third function for now. But I do expect that you will find it to be almost identical to the existing transparent one.

Not exactly. Because if the Host is missing the "prepareTransparentURL" uses the  ConnStateData::clientConnection->local to build url. This is OK for transparent SSL bumping but not when we are bump normal CONNECT requests.

chtsanti commented:
I added the ConnStateData::prepareTlsSwitchingURL to generate URLs for ssl bumped requests. It is ConnStateData because it uses private members of ConnStateData.
I add the  ConnStateData::tlsConnectPort to hold the port on remote server for bumped requests. It is the only way to has port for all the cases.
I also changed the prepareTransparentURL and prepareAcceleratedURL functions. PLease see if you agree with the changes.

yadij commented on src/client_side.cc:
```diff
@@ -1162,67 +1158,116 @@ prepareAcceleratedURL(ConnStateData * conn, ClientHttpRequest *http, const Http1
-        debugs(33, 5, "TRANSPARENT HOST REWRITE: " << http->uri);
-    } else {
+        const int url_sz = scheme.length() + strlen(host) + hp->requestUri().length() + 32;
+        uri = (char *)xcalloc(url_sz, 1);
```
casting

yadij commented on src/client_side.cc:
```diff
@@ -1162,67 +1158,116 @@ prepareAcceleratedURL(ConnStateData * conn, ClientHttpRequest *http, const Http1
         const SBuf &scheme = AnyP::UriScheme(conn->transferProtocol.protocol).image();
-        snprintf(http->uri, url_sz, SQUIDSBUFPH "://%s:%d" SQUIDSBUFPH,
+        const int url_sz = scheme.length() + sizeof(ipbuf) + url.length() + 32;
+        char *uri = (char *)xcalloc(url_sz, 1);
```
casting

yadij commented on src/client_side.cc:
```diff
@@ -1162,67 +1158,116 @@ prepareAcceleratedURL(ConnStateData * conn, ClientHttpRequest *http, const Http1
         const SBuf &scheme = AnyP::UriScheme(conn->transferProtocol.protocol).image();
-        snprintf(http->uri, url_sz, SQUIDSBUFPH "://%s%s" SQUIDSBUFPH,
+        const int url_sz = scheme.length() + strlen(conn->port->defaultsite) + sizeof(vportStr) + url.length() + 32;
+        char *uri = (char *)xcalloc(url_sz, 1);
```
casting

yadij commented on src/client_side.cc:
```diff
@@ -1162,67 +1158,116 @@ prepareAcceleratedURL(ConnStateData * conn, ClientHttpRequest *http, const Http1
-        snprintf(http->uri, url_sz, SQUIDSBUFPH "://%s" SQUIDSBUFPH, SQUIDSBUFPRINT(scheme), host, SQUIDSBUFPRINT(url));
-        debugs(33, 5, "ACCEL VHOST REWRITE: " << http->uri);
+        const int url_sz = scheme.length() + strlen(host) + url.length() + 32;
+        char *uri = (char *)xcalloc(url_sz, 1);
```
casting

yadij commented on src/client_side.cc:
```diff
@@ -1162,67 +1158,116 @@ prepareAcceleratedURL(ConnStateData * conn, ClientHttpRequest *http, const Http1
+
+        const SBuf &scheme = AnyP::UriScheme(transferProtocol.protocol).image();
+        const int url_sz = scheme.length() + useHost.length() + hp->requestUri().length() + 32;
+        uri = (char *)xcalloc(url_sz, 1);
```
casting

rousskov commented:
@eduard-bagdasaryan, please check why Anubis is not updating this (or the next `M-cleared-for-merge`) PR. I am probably missing something, but it feels like Anubis is stuck.

eduard-bagdasaryan commented:
> it feels like Anubis is stuck

It got stuck on PR30. The reason is unclear, I suspect it hung during an API call of getting statuses from GitHub. So I have to I restarted it, and now PR306 is being processed.

rousskov commented:
Glad I asked! Thank you for addressing the immediate problem. Long-term, does Anubis need API call timeouts? I thought we already had those...

eduard-bagdasaryan commented:
> It got stuck on PR30.

I found the true reason of this problem and fixed.

chtsanti commented on src/client_side.cc:
```diff
@@ -1162,67 +1158,116 @@ prepareAcceleratedURL(ConnStateData * conn, ClientHttpRequest *http, const Http1
+        return nullptr; /* already in good shape */
+
+    char *uri = buildUrlFromHost(conn, hp);
+    if (!uri){
```
ok

chtsanti commented on src/client_side.cc:
```diff
@@ -1162,67 +1158,116 @@ prepareAcceleratedURL(ConnStateData * conn, ClientHttpRequest *http, const Http1
+        conn->clientConnection->local.toHostStr(ipbuf,MAX_IPSTRLEN);
+        const SBuf &scheme = AnyP::UriScheme(conn->transferProtocol.protocol).image();
+        const int url_sz = sizeof(ipbuf) + hp->requestUri().length() + 32;
+        uri = (char *)xcalloc(url_sz, 1);
```
ok

chtsanti commented on src/client_side.cc:
```diff
@@ -1142,18 +1140,16 @@ prepareAcceleratedURL(ConnStateData * conn, ClientHttpRequest *http, const Http1
             thost[0] = '\0';
             char *t = NULL;
-            if (host[strlen(host)] != ']' && (t = strrchr(host,':')) != NULL) {
+            if (host[strlen(host) - 1] != ']' && (t = strrchr(host,':')) != NULL) {
```
ok

yadij commented on src/client_side.h:
```diff
@@ -397,6 +398,7 @@ class ConnStateData : public Server, public HttpControlMsgSink, private Independ
 
     /// The SSL server host name appears in CONNECT request or the server ip address for the intercepted requests
     String sslConnectHostOrIp; ///< The SSL server host name as passed in the CONNECT request
+    unsigned short tlsConnectPort; ///< The TLS server port name as passed in the CONNECT request
```
number
```suggestion
    unsigned short tlsConnectPort; ///< The TLS server port number as passed in the CONNECT request
```

chtsanti replied:
addressed at 19f00737

yadij commented on src/client_side.cc:
```diff
@@ -3051,6 +3100,7 @@ ConnStateData::switchToHttps(HttpRequest *request, Ssl::BumpMode bumpServerMode)
     assert(!switchedToHttps_);
 
     sslConnectHostOrIp = request->url.host();
+    tlsConnectPort =  request->url.port();
```
whitespace
```suggestion
    tlsConnectPort = request->url.port();
```

chtsanti replied:
addressed at 19f00737

yadij commented on src/client_side.cc:
```diff
@@ -1342,9 +1387,11 @@ parseHttpRequest(ConnStateData *csd, const Http1::RequestParserPointer &hp)
-    if (csd->transparent()) {
+    if (csd->switchedToHttps()) {
+        http->uri = csd->prepareTlsSwitchingURL(hp);
+    } else if (csd->transparent() || csd->switchedToHttps()) {
```
This switchedToHttps() condition can never be true, if it was the if-statement above would have run instead.
```suggestion
    } else if (csd->transparent()) {
```

chtsanti replied:
addressed at 19f00737

chtsanti commented:
@yadij 
I have addressed all change requests in your review. Please re-review.

yadij commented:
Updating to master to trigger the Jenkins re-build. Earlier failures were node errors, not code related.
